src/token/loginack-token-parser.ts

```diff
@@ -1,46 +1,73 @@
-import Parser, { ParserOptions } from './stream-parser';
-
+import LegacyParser, { ParserOptions } from './stream-parser';
 import { LoginAckToken } from './token';
 
+import { UInt8, UInt32BE, UsVarbyte, BVarchar, Map, Record } from '../parser';
+
 import { versionsByValue as versions } from '../tds-versions';
 
 const interfaceTypes: { [key: number]: string } = {
   0: 'SQL_DFLT',
   1: 'SQL_TSQL'
 };
 
-function loginAckParser(parser: Parser, _options: ParserOptions, callback: (token: LoginAckToken) => void) {
-  // length
-  parser.readUInt16LE(() => {
-    parser.readUInt8((interfaceNumber) => {
-      const interfaceType = interfaceTypes[interfaceNumber];
-      parser.readUInt32BE((tdsVersionNumber) => {
-        const tdsVersion = versions[tdsVersionNumber];
-        parser.readBVarChar((progName) => {
-          parser.readUInt8((major) => {
-            parser.readUInt8((minor) => {
-              parser.readUInt8((buildNumHi) => {
-                parser.readUInt8((buildNumLow) => {
-                  callback(new LoginAckToken({
-                    interface: interfaceType,
-                    tdsVersion: tdsVersion,
-                    progName: progName,
-                    progVersion: {
-                      major: major,
-                      minor: minor,
-                      buildNumHi: buildNumHi,
-                      buildNumLow: buildNumLow
-                    }
-                  }));
-                });
-              });
-            });
-          });
-        });
-      });
+interface TokenData {
+  interfaceNumber: number;
+  tdsVersionNumber: number;
+  progName: string;
+  major: number;
+  minor: number;
+  buildNumHi: number;
+  buildNumLow: number;
+}
+
+class TokenDataParser extends Record<TokenData> {
+  constructor() {
+    super({
+      interfaceNumber: new UInt8(),
+      tdsVersionNumber: new UInt32BE(),
+      progName: new BVarchar(),
+      major: new UInt8(),
+      minor: new UInt8(),
+      buildNumHi: new UInt8(),
+      buildNumLow: new UInt8(),
```

Comment:
Because the parsers are combinable, you can do this:

```suggestion
  constructor() {
    super({
      interface: new Map(new UInt8(), (interfaceNumber) => {
        return interfaceTypes[data.interfaceNumber];
      }),
      tdsVersion: new Map(new UInt32BE(), (tdsVersionNumber) => {
        return versions[data.tdsVersionNumber];
      }),
      progName: new BVarchar(),
      progVersion: new Record({
        major: new UInt8(),
        minor: new UInt8(),
        buildNumHi: new UInt8(),
        buildNumLow: new UInt8(),
      })
```